### ISSUE TITLE
Error when trying to launch from IntelliJ as External Tool

### ISSUE THREAD

**Issue by Demozo** [ORelio/Minecraft-Console-Client#603]
When added as an [External Tool](https://www.jetbrains.com/help/idea/settings-tools-external-tools.html) in IntelliJ I get the following error.

> Cannot read keys when either application does not have a console or when console input has been redirected from a file. Try Console.Read.

This seems to be caused by the program/C# not seeing IntelliJ as an actual console but as a file input.
I'm trying to run it like `C:\Path\To\The\Exe\MinecraftClient.exe "" "" localhost` with IntelliJ setting the working dir to `C:\Path\To\The\Exe\`

I'll give it a look to see if I can fix it myself and submit a PR. If not it would be greatly appreciated if someone else could do so.

**Comment by TheSnoozer** [comment]
IntelliJ IDEA is a development environment for developing computer software for Java. 

Based on the link you send:
> [External tools] define third-party standalone applications (code generators and analyzers, pre- and post- processors, database utilities, etc.) as external tools to be able to run them from IntelliJ IDEA. 

This project provides you a minecraft console client that is written in C# (a different language). It furthermore does **not** provide a standalone application that would fit the description for an external tool (e.g. this project does **not** provide code generators, analyzers, pre- and post- processors, database utilities, or anything related).

In general you launch the client by running it from the Windows command line or using mono that allows you to also launch it on linux and mac from the terminal.

AFAIK there is no reason and furthermore absolutely no benefit to run the client as an external tool from IntelliJ IDEA, since it simply does not fit the purpose of an external tool. 

So why would you want to run the minecraft console client with IntelliJ IDEA as external tool?

**Comment by Demozo** [comment]
I am very well aware of that, I'm a programmer for a living.

IntelliJ simply launches the program with arguments. And by default also opens a console window in the IDE showing output from the tool/program run in question. Console.ReadKey is used in the GetPassword method of the client. Meaning that it does run, but encounters an error in that specific method. The username section most likely does use Console.Read.

The benefit of running this through IntelliJ is that the output of the client will appear in the editor as opposed to having its own window, thus removing the necessity to alt tab.

Regardless, this is most likely not even an IntelliJ specific issue but an issue that will be produced by anything that re-routes the stdin stream as it will be seen as reading from a file by C#.

Apologies for the lack of proper formatting, I'm typing this from my phone.

**Comment by Pokechu22** [comment]
If I recall correctly, you can enable basic IO mode in the configuration where it'll no longer use GetPassword.

**Comment by ORelio** [comment]
If you cannot provide an interactive terminal you can switch to basic IO mode by specifying the undocumented `BasicIO` string (case sensitive) as last argument. Chat color and some other things such as interactive chat prompt will be disabled so the output will be quite messed up but it should work with redirected I/O.

**Comment by TheSnoozer** [comment]
Could you get this sorted?

**Comment by Demozo** [comment]
Switching `Console.ReadKey(true)` to `Console.Read()` does several things, none of which are positive.
* It removes the ability to mask the password with asterisks
* It complicates the switch statement that's currently in place
* It is very edge-case prone, and fixing those would afaik require writing disgusting code

BasicIO does indeed work in IntelliJ, but as it doesn't use the `ReadKey` method, it can't mask the password. Output isn't as "messed up" as @ORelio seemed to indicate, so that's good. Afaik it's just the occasional paragraph symbol thrown into the output as it can't parse colors.

So in conclusion I think this issue is with the way `ReadKey` works in C#, and as such can't be fixed. I am really curious about the reasoning behind why reading from a "file" aka rerouted input stream doesn't work with `ReadKey`.

**Comment by ORelio** [comment]
@TheSnoozer No, this cannot be fixed without removing the I/O prompt which requires an actual terminal. `BasicIO` is designed to use MCC from another program (not a terminal), that should match your use case.
@Demozo The only mess-up that is likely to occur is if some chat message is received while you are typing. If you never type anything, the only major difference is color codes being printed as `§c` instead of setting the terminal color (which does not make sense when not using a terminal).
